plugin/FERC/render.py

```diff
@@ -860,30 +860,39 @@ def substitute_rule(rule_name, sub_info, line_number_subs, rule_results, templat
                         rule_focus_index = get_rule_focus_index(all_result_part or json_rule_result, json_result)
                         if rule_focus_index is not None:
                             use_refs = refs or rule_result.refs
-                            fact_object_index = use_refs[rule_focus_index]['objectId']
-                            model_fact = modelXbrl.modelObject(fact_object_index)
-                            processed_facts.add(model_fact)
-                            expression_node = get_node_by_pos(template, sub['expression-node'])
-                            content, new_fact_id = format_fact(expression_node, 
-                                                               model_fact, 
-                                                               main_html, 
-                                                               # This will check if the html flag is in meta data (sub) or calculated in the result
-                                                               sub.get('html', json_result.get('html', False)), 
-                                                               json_result, 
-                                                               template_nsmap, 
-                                                               fact_number)
-
-                                            
-                            # Save the context and unit ids
-                            context_ids.add(model_fact.contextID)
-                            if model_fact.unitID is not None:
-                                unit_ids.add(model_fact.unitID)
-                            # Check if the fact is redacted
-                            is_redacted = fact_is_marked(model_fact, 'http://www.ferc.gov/arcrole/Redacted')
-                            # Check if the fact is confidential
-                            is_confidential = fact_is_marked(model_fact, 'http://www.ferc.gov/arcrole/Confidential')                                
-                            # Check if there are footnotes
-                            current_footnote_ids = get_footnotes(footnotes, model_fact, sub, new_fact_id, is_confidential, is_redacted)
+                            fact_object_index = use_refs[rule_focus_index]['objectId'] if rule_focus_index < len(use_refs) else None
+                            if fact_object_index is not None:
```

Comment:
We encountered this with an erroneous FERC instance document when processing form1 t70-xule-text.xule.  The end of that rule looks like this:
```
for $row in $uniqueData.sort
list(dict(list('type', 'f'), list('part', 0), list('value', (if exists($row[2]) (($row[2])#rv-0).string else (none)#rv-0).string), list('is-fact', if exists($row[2]) (($row[2]).is-fact).string else 'false')),
dict(list('type', 'f'), list('part', 1), list('value', (if exists($row[3]) (($row[3])#rv-1).string else (none)#rv-1).string), list('is-fact', if exists($row[3]) (($row[3]).is-fact).string else 'false')),
dict(list('type', 'f'), list('part', 2), list('value', (if exists($row[4]) (($row[4])#rv-2).string else (none)#rv-2).string), list('is-fact', if exists($row[4]) (($row[4]).is-fact).string else 'false')),
dict(list('type', 'f'), list('part', 3), list('value', (if exists($row[5]) (($row[5])#rv-3).string else (none)#rv-3).string), list('is-fact', if exists($row[5]) (($row[5]).is-fact).string else 'false')),
dict(list('type', 'f'), list('part', 4), list('value', (if exists($row[6]) (($row[6])#rv-4).string else (none)#rv-4).string), list('is-fact', if exists($row[6]) (($row[6]).is-fact).string else 'false')),
dict(list('type', 'f'), list('part', 5), list('value', (if exists($row[7]) (($row[7])#rv-5).string else (none)#rv-5).string), list('is-fact', if exists($row[7]) (($row[7]).is-fact).string else 'false')),
dict(list('type', 'f'), list('part', 6), list('value', (if exists($row[8]) (($row[8])#rv-6).string else (none)#rv-6).string), list('is-fact', if exists($row[8]) (($row[8]).is-fact).string else 'false'))).to-json
rule-focus list(if $rv-0.is-fact $rv-0 else none, if $rv-1.is-fact $rv-1 else none, if $rv-2.is-fact $rv-2 else none, if $rv-3.is-fact $rv-3 else none, if $rv-4.is-fact $rv-4 else none, if $rv-5.is-fact $rv-5 else none, if $rv-6.is-fact $rv-6 else none)
```
It seems as if the intent of the generated XULE + Python code is to try to capture "all" fact objects used by the expression, and the `'is-fact'` key in the result objects is used to help skip over None in the list that is filtered out by `ModelXbrl.log`, which is seemingly used by `evaluate_output_rule` in order to convert objects to ids.  However, if the expression ends in `for`, then each row of the result is matched against the rule-focus, but the rule-focus can seemingly only capture one row of output, so it seems as if that matching does't always work reliably.  In the specific error case we observed, there were more facts in the row than in the captured rule-focus row, so the indexing fails.

The change here is a workaround for that deeper issue, but it was sufficient for our users to be able to diagnose the issue, so we're posting it for illustration.  A proper fix probably needs to somehow realign the output to the rule-focus, but we're not quite sure the best way to do that (and in any case, it's not yet clear to us how to rebuild/regenerate the renderer XULE files).